In [1]:
import pandas as pd
import numpy as np

#### Dataset: Amazon Prime TV shows and titles.

- The goal of this project is exploratory data analysis to potentially derive interesting and valuable insights about Amazon TV shows and titles.
- Two (2) csv files are available for use in this project and  are derived from the Kaggle (link is provided below)
- Source of data: https://www.kaggle.com/datasets/victorsoeiro/amazon-prime-tv-shows-and-movies?resource=download&select=titles.csv

In [2]:
# Load the csv files Amazon Prime TV shows
tv_credits = pd.read_csv(r'C:\Users\LuisV\Desktop\UDACITY\Portfolio\Amazon Prime TV shows and titles\amazon_project\credits.csv')

In [3]:
# Display the first 5 rows of the dataframe
tv_credits.head(5)

person_id       id           name      character   role
0      59401  ts20945     Joe Besser            Joe  ACTOR
1      31460  ts20945     Moe Howard            Moe  ACTOR
2      31461  ts20945     Larry Fine          Larry  ACTOR
3      21174  tm19248  Buster Keaton    Johnny Gray  ACTOR
4      28713  tm19248    Marion Mack  Annabelle Lee  ACTOR

In [4]:
# Display the number of rows and columns in the dataframe
tv_credits.shape

(124235, 5)

In [5]:
# Create a copy of the dataframe
tv = tv_credits.copy()

In [6]:
# Check for the number of rows and columns in the copy of the dataframe
tv.shape

(124235, 5)

In [7]:
# Check for the total number of duplicate rows in the data
tv.duplicated().sum()

56

- There are 56 rows duplicate rows in the dataframe

In [8]:
# Drop all duplicate rows across all columns
tv = tv.drop_duplicates()

In [9]:
# Check for the number of rows and columns after removing the duplicates
tv.shape

(124179, 5)

- There are 124,179 rows left after removal of duplicate rows across all columns

In [10]:
# Check for the total number of duplicate rows in the dataframe
tv.duplicated().sum()

0

In [11]:
# Check for the number of null values in the dataframe
tv.isnull().sum()

person_id        0
id               0
name             0
character    16277
role             0
dtype: int64

- There are 16,277 empty rows in the Character column, while other columns have no empty entry.

In [12]:
# Fill missing values in the Character column with a forward fill method
tv = tv.fillna(method = 'ffill', axis = 0)

In [13]:
# Check for the number of null values in the dataframe
tv.isnull().sum()

person_id    0
id           0
name         0
character    0
role         0
dtype: int64

- There are no null values left in the dataframe

In [14]:
# Load the csv files Amazon Prime TV shows
titles = pd.read_csv(r'C:\Users\LuisV\Desktop\UDACITY\Portfolio\Amazon Prime TV shows and titles\amazon_project\titles.csv')

In [15]:
# Check to see the number of rows and columns in the dataframe
titles.shape

(9871, 15)

In [16]:
#Check the first few rows of the dataframe for movie titles
titles.head(5)

id                        title   type  \
0  ts20945            The Three Stooges   SHOW   
1  tm19248                  The General  MOVIE   
2  tm82253  The Best Years of Our Lives  MOVIE   
3  tm83884              His Girl Friday  MOVIE   
4  tm56584            In a Lonely Place  MOVIE   

                                         description  release_year  \
0  The Three Stooges were an American vaudeville ...          1934   
1  During America’s Civil War, Union spies steal ...          1926   
2  It's the hope that sustains the spirit of ever...          1946   
3  Hildy, the journalist former wife of newspaper...          1940   
4  An aspiring actress begins to suspect that her...          1950   

  age_certification  runtime  \
0             TV-PG       19   
1               NaN       78   
2               NaN      171   
3               NaN       92   
4               NaN       94   

                                              genres production_countries  \
0  ['comedy', 'family', 'animation', 'action', 'f...               ['US']   
1  ['action', 'drama', 'war', 'western', 'comedy'...               ['US']   
2                        ['romance', 'war', 'drama']               ['US']   
3                     ['comedy', 'drama', 'romance']               ['US']   
4                   ['thriller', 'drama', 'romance']               ['US']   

   seasons    imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0     26.0  tt0850645         8.6      1092.0           15.424         7.6  
1      NaN  tt0017925         8.2     89766.0            8.647         8.0  
2      NaN  tt0036868         8.1     63026.0            8.435         7.8  
3      NaN  tt0032599         7.8     57835.0           11.270         7.4  
4      NaN  tt0042593         7.9     30924.0            8.273         7.6

In [17]:
# Print the summary of the dataframe
titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9871 entries, 0 to 9870
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    9871 non-null   object 
 1   title                 9871 non-null   object 
 2   type                  9871 non-null   object 
 3   description           9752 non-null   object 
 4   release_year          9871 non-null   int64  
 5   age_certification     3384 non-null   object 
 6   runtime               9871 non-null   int64  
 7   genres                9871 non-null   object 
 8   production_countries  9871 non-null   object 
 9   seasons               1357 non-null   float64
 10  imdb_id               9204 non-null   object 
 11  imdb_score            8850 non-null   float64
 12  imdb_votes            8840 non-null   float64
 13  tmdb_popularity       9324 non-null   float64
 14  tmdb_score            7789 non-null   float64
dtypes: float64(5), int64(

In [18]:
# Check for the duplicates in the dataframe
titles.duplicated().sum()

3

- There are 3 duplicate rows in the dataframe

In [19]:
# Drop the duplicate rows in the movie titles dataframe
titles = titles.drop_duplicates()

In [20]:
# Check for the number of duplicates in the dataframe
titles.duplicated().sum()

0

- Duplicate rows have been removed

In [21]:
# Check that there are now 9,868 rows instead of 9,871 since duplicates have been removed
titles.shape

(9868, 15)

In [22]:
# Display the total number of null values for each column in the dataframe
titles.isnull().sum()

id                         0
title                      0
type                       0
description              119
release_year               0
age_certification       6484
runtime                    0
genres                     0
production_countries       0
seasons                 8511
imdb_id                  667
imdb_score              1021
imdb_votes              1031
tmdb_popularity          547
tmdb_score              2080
dtype: int64

- For the seasons column, over 86% of the data is missing and therefore the column will be dropped.

In [23]:
#create a copy of the dataframe
title = titles.copy()

In [24]:
# Check the first few rows of the dataframe
title.head(5)

id                        title   type  \
0  ts20945            The Three Stooges   SHOW   
1  tm19248                  The General  MOVIE   
2  tm82253  The Best Years of Our Lives  MOVIE   
3  tm83884              His Girl Friday  MOVIE   
4  tm56584            In a Lonely Place  MOVIE   

                                         description  release_year  \
0  The Three Stooges were an American vaudeville ...          1934   
1  During America’s Civil War, Union spies steal ...          1926   
2  It's the hope that sustains the spirit of ever...          1946   
3  Hildy, the journalist former wife of newspaper...          1940   
4  An aspiring actress begins to suspect that her...          1950   

  age_certification  runtime  \
0             TV-PG       19   
1               NaN       78   
2               NaN      171   
3               NaN       92   
4               NaN       94   

                                              genres production_countries  \
0  ['comedy', 'family', 'animation', 'action', 'f...               ['US']   
1  ['action', 'drama', 'war', 'western', 'comedy'...               ['US']   
2                        ['romance', 'war', 'drama']               ['US']   
3                     ['comedy', 'drama', 'romance']               ['US']   
4                   ['thriller', 'drama', 'romance']               ['US']   

   seasons    imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0     26.0  tt0850645         8.6      1092.0           15.424         7.6  
1      NaN  tt0017925         8.2     89766.0            8.647         8.0  
2      NaN  tt0036868         8.1     63026.0            8.435         7.8  
3      NaN  tt0032599         7.8     57835.0           11.270         7.4  
4      NaN  tt0042593         7.9     30924.0            8.273         7.6

In [25]:
# Check the number of rows and columns in the copy of the dataframe
title.shape

(9868, 15)

In [26]:
# Drop the seasons column from the dataframe and save the result as a new dataframe
no_seasons = title.drop(['seasons'], axis=1)

In [27]:
no_seasons.tail(5)

id                            title   type  \
9866   tm510327                     Lily Is Here  MOVIE   
9867  tm1079144  Jay Nog: Something from Nothing  MOVIE   
9868   tm847725                          Chasing  MOVIE   
9869  tm1054116                         Baikunth  MOVIE   
9870   ts275838                 Waking Up Eighty   SHOW   

                                            description  release_year  \
9866  Dallas and heroin have one thing in common: Du...          2021   
9867  Something From Nothing takes you on a stand-up...          2021   
9868  A cop from Chennai sets out to nab a dreaded d...          2021   
9869  This story is about prevalent caste problem, e...          2021   
9870  Kara Stewart, 16, is fed up with just about ev...          2021   

     age_certification  runtime               genres production_countries  \
9866               NaN       93            ['drama']               ['US']   
9867               NaN       55           ['comedy']               ['US']   
9868               NaN      116            ['crime']               ['IN']   
9869               NaN       72  ['family', 'drama']                   []   
9870               NaN       10            ['drama']                   []   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
9866   tt7672388         5.3        20.0            1.406         NaN  
9867  tt15041600         NaN         NaN            0.600         NaN  
9868         NaN         NaN         NaN            1.960         NaN  
9869  tt14331982         8.4        49.0            0.645         NaN  
9870  tt13542552         NaN         NaN              NaN         NaN

- The age certification column contains 6,484 null values.This is ~66% of missing data so the column will be dropped.

In [28]:
# Drop the age_certification and imdb_id columns from the dataframe. An 'id' column with unique values already exists
# in the dataframe so the imdb_id column does not necessarily serve as an important column for exploratory data analysis
no_age = no_seasons.drop(['age_certification','imdb_id'], axis=1)   

In [29]:
# Check the first few rows of the dataframe
no_age.head(5)

id                        title   type  \
0  ts20945            The Three Stooges   SHOW   
1  tm19248                  The General  MOVIE   
2  tm82253  The Best Years of Our Lives  MOVIE   
3  tm83884              His Girl Friday  MOVIE   
4  tm56584            In a Lonely Place  MOVIE   

                                         description  release_year  runtime  \
0  The Three Stooges were an American vaudeville ...          1934       19   
1  During America’s Civil War, Union spies steal ...          1926       78   
2  It's the hope that sustains the spirit of ever...          1946      171   
3  Hildy, the journalist former wife of newspaper...          1940       92   
4  An aspiring actress begins to suspect that her...          1950       94   

                                              genres production_countries  \
0  ['comedy', 'family', 'animation', 'action', 'f...               ['US']   
1  ['action', 'drama', 'war', 'western', 'comedy'...               ['US']   
2                        ['romance', 'war', 'drama']               ['US']   
3                     ['comedy', 'drama', 'romance']               ['US']   
4                   ['thriller', 'drama', 'romance']               ['US']   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.6      1092.0           15.424         7.6  
1         8.2     89766.0            8.647         8.0  
2         8.1     63026.0            8.435         7.8  
3         7.8     57835.0           11.270         7.4  
4         7.9     30924.0            8.273         7.6

In [30]:
# Display the total number of null values left for each column in the dataframe
no_age.isnull().sum()

id                         0
title                      0
type                       0
description              119
release_year               0
runtime                    0
genres                     0
production_countries       0
imdb_score              1021
imdb_votes              1031
tmdb_popularity          547
tmdb_score              2080
dtype: int64

In [31]:
# Check the last 5 rows of the dataframe
no_age.tail(5)

id                            title   type  \
9866   tm510327                     Lily Is Here  MOVIE   
9867  tm1079144  Jay Nog: Something from Nothing  MOVIE   
9868   tm847725                          Chasing  MOVIE   
9869  tm1054116                         Baikunth  MOVIE   
9870   ts275838                 Waking Up Eighty   SHOW   

                                            description  release_year  \
9866  Dallas and heroin have one thing in common: Du...          2021   
9867  Something From Nothing takes you on a stand-up...          2021   
9868  A cop from Chennai sets out to nab a dreaded d...          2021   
9869  This story is about prevalent caste problem, e...          2021   
9870  Kara Stewart, 16, is fed up with just about ev...          2021   

      runtime               genres production_countries  imdb_score  \
9866       93            ['drama']               ['US']         5.3   
9867       55           ['comedy']               ['US']         NaN   
9868      116            ['crime']               ['IN']         NaN   
9869       72  ['family', 'drama']                   []         8.4   
9870       10            ['drama']                   []         NaN   

      imdb_votes  tmdb_popularity  tmdb_score  
9866        20.0            1.406         NaN  
9867         NaN            0.600         NaN  
9868         NaN            1.960         NaN  
9869        49.0            0.645         NaN  
9870         NaN              NaN         NaN

In [32]:
no_age.shape

(9868, 12)

In [33]:
# Replace null values in the following columns using a forward_filling method: 'tmdb_score','imdb_score', and'imdb_votes'
no_age[['tmdb_score','imdb_score','imdb_votes']]= no_age[['tmdb_score','imdb_score','imdb_votes']].ffill()

In [34]:
# Display the total number of null values left for each column in the dataframe
no_age.isnull().sum()

id                        0
title                     0
type                      0
description             119
release_year              0
runtime                   0
genres                    0
production_countries      0
imdb_score                0
imdb_votes                0
tmdb_popularity         547
tmdb_score                0
dtype: int64

In [35]:
# Replace the null values in the 'tmdb_popularity' column with 0

no_age['tmdb_popularity'] = no_age['tmdb_popularity'].fillna(0)

In [36]:
# Display the total number of null values per column
no_age.isnull().sum()

id                        0
title                     0
type                      0
description             119
release_year              0
runtime                   0
genres                    0
production_countries      0
imdb_score                0
imdb_votes                0
tmdb_popularity           0
tmdb_score                0
dtype: int64

In [37]:
# Use a forward fill method to replace the null values in the description column
no_age['description']= no_age['description'].ffill()

In [53]:
# Display the total number of null values to check if any column has null values left
no_age.isnull().sum()

id                      0
title                   0
type                    0
description             0
release_year            0
runtime                 0
genres                  0
production_countries    0
imdb_score              0
imdb_votes              0
tmdb_popularity         0
tmdb_score              0
dtype: int64

- Dataframe is clean

In [54]:
#Create a copy of the cleaned dataframe
movie_titles_data = no_age.copy()

In [55]:
# Check the first 10 rows of the dataframe
movie_titles_data.head(5)

id                        title   type  \
0  ts20945            The Three Stooges   SHOW   
1  tm19248                  The General  MOVIE   
2  tm82253  The Best Years of Our Lives  MOVIE   
3  tm83884              His Girl Friday  MOVIE   
4  tm56584            In a Lonely Place  MOVIE   

                                         description  release_year  runtime  \
0  The Three Stooges were an American vaudeville ...          1934       19   
1  During America’s Civil War, Union spies steal ...          1926       78   
2  It's the hope that sustains the spirit of ever...          1946      171   
3  Hildy, the journalist former wife of newspaper...          1940       92   
4  An aspiring actress begins to suspect that her...          1950       94   

                                              genres production_countries  \
0  ['comedy', 'family', 'animation', 'action', 'f...               ['US']   
1  ['action', 'drama', 'war', 'western', 'comedy'...               ['US']   
2                        ['romance', 'war', 'drama']               ['US']   
3                     ['comedy', 'drama', 'romance']               ['US']   
4                   ['thriller', 'drama', 'romance']               ['US']   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.6      1092.0           15.424         7.6  
1         8.2     89766.0            8.647         8.0  
2         8.1     63026.0            8.435         7.8  
3         7.8     57835.0           11.270         7.4  
4         7.9     30924.0            8.273         7.6

In [68]:
# Remove square brackets from string value of production countries
movie_titles_data['production_countries'] = movie_titles_data['production_countries'].str.replace("[","").str.replace("]","")

C:\Users\LuisV\AppData\Local\Temp/ipykernel_11036/1002774130.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  movie_titles_data['production_countries'] = movie_titles_data['production_countries'].str.replace("[","").str.replace("]","")


In [72]:
# Remove single quotes from string value of production countries
movie_titles_data['production_countries'] = movie_titles_data['production_countries'].str.replace("'","").str.replace("'","")

In [73]:
# Check the first 3 rows of the dataframe to make sure the square brackets and single quotes 
#have been removed from string values in the production_countries column
movie_titles_data.head(3)

id                        title   type  \
0  ts20945            The Three Stooges   SHOW   
1  tm19248                  The General  MOVIE   
2  tm82253  The Best Years of Our Lives  MOVIE   

                                         description  release_year  runtime  \
0  The Three Stooges were an American vaudeville ...          1934       19   
1  During America’s Civil War, Union spies steal ...          1926       78   
2  It's the hope that sustains the spirit of ever...          1946      171   

                                              genres production_countries  \
0  ['comedy', 'family', 'animation', 'action', 'f...                   US   
1  ['action', 'drama', 'war', 'western', 'comedy'...                   US   
2                        ['romance', 'war', 'drama']                   US   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.6      1092.0           15.424         7.6  
1         8.2     89766.0            8.647         8.0  
2         8.1     63026.0            8.435         7.8

### Exploratory Data Analysis of the movie titles dataframe

- The main goal is to explore as much of the movie titles data to uncover interesting insights.

In [84]:
# Find the total number of different categories of titles in the dataframe
counts = movie_titles_data['type'].nunique()

In [85]:

counts

2

- There are only 2 different categories of titles 

In [86]:
# Count the total number of each category type
num = movie_titles_data['type'].value_counts()

In [87]:
num


MOVIE    8511
SHOW     1357
Name: type, dtype: int64

- There are 1,357 shows and 8,511 movie titles

In [80]:
# Display data for TV shows only
shows_only = movie_titles_data[movie_titles_data["type"] == 'SHOW']


In [81]:
# Display on the fist 5 rows of the dataframe
shows_only.head(5)

id                        title  type  \
0    ts20945            The Three Stooges  SHOW   
91   ts55748              What's My Line?  SHOW   
956  ts20005                  I Love Lucy  SHOW   
969  ts42867  Mister Rogers' Neighborhood  SHOW   
980  ts21930              Lupin the Third  SHOW   

                                           description  release_year  runtime  \
0    The Three Stooges were an American vaudeville ...          1934       19   
91   Four panelists must determine guests' occupati...          1950       30   
956  Cuban Bandleader Ricky Ricardo would be happy ...          1951       30   
969  Mister Rogers' Neighborhood is an American chi...          1968       29   
980  Arsene Lupin III is the grandson of the master...          1971       23   

                                                genres production_countries  \
0    ['comedy', 'family', 'animation', 'action', 'f...                   US   
91                               ['reality', 'family']                   US   
956                               ['comedy', 'family']                   US   
969                     ['fantasy', 'music', 'family']                   US   
980  ['scifi', 'animation', 'action', 'comedy', 'cr...                   JP   

     imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0           8.6      1092.0           15.424         7.6  
91          8.6      1563.0           87.392         6.9  
956         8.5     25944.0           17.088         8.1  
969         8.7      8675.0            8.747         4.7  
980         7.9      2116.0           45.829         8.0

In [ ]:
# Display the top 10 shows based on the movie database popularity score (i.e. tmdb_popularity)

In [76]:
# Sort the dataframe by the tmdb_popularity (the movie database) column in a descending order 
# and limit result to the first 10 rows.This will reveal the top 10 most popular movies in the movie database 

movie_titles_data.sort_values(by='tmdb_popularity', ascending=False).head(10)

id                              title   type  \
8934   tm266536                 All the Old Knives  MOVIE   
9040   ts331561                             Harina   SHOW   
9039  tm1188007                  The eighth clause  MOVIE   
8964   tm460828  Hotel Transylvania: Transformania  MOVIE   
7421   tm374139                 Sonic the Hedgehog  MOVIE   
8948   tm460993           Clifford the Big Red Dog  MOVIE   
9092  tm1042364                    Queen of Spades  MOVIE   
9011   tm943374                            Meander  MOVIE   
4099    ts22277                              Suits   SHOW   
5287        ts8                   Better Call Saul   SHOW   

                                            description  release_year  \
8934  When the CIA discovers one of its agents leake...          2022   
9040  The misadventures of Lieutenant Harina and his...          2022   
9039  Kat and Borja appear to be a perfect couple, b...          2022   
8964  When Van Helsing's mysterious invention, the "...          2022   
7421  Powered with incredible speed, Sonic The Hedge...          2020   
8948  As Emily struggles to fit in at home and at sc...          2021   
9092  According to legend, an ominous entity known a...          2021   
9011  After getting a car ride from an unknown man, ...          2021   
4099  While running from a drug deal gone bad, Mike ...          2011   
5287  Six years before Saul Goodman meets Walter Whi...          2015   

      runtime                                             genres  \
8934      101                                       ['thriller']   
9040       33                                         ['comedy']   
9039       80                                       ['thriller']   
8964       87  ['fantasy', 'romance', 'animation', 'comedy', ...   
7421       99  ['scifi', 'action', 'comedy', 'family', 'fanta...   
8948       97  ['fantasy', 'comedy', 'family', 'animation', '...   
9092       91                                         ['horror']   
9011       91           ['horror', 'thriller', 'drama', 'scifi']   
4099       44                                ['drama', 'comedy']   
5287       49                       ['comedy', 'drama', 'crime']   

     production_countries  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
8934                   US         6.1     14686.0         1437.906         6.0  
9040                   MX         7.8       509.0          951.863         5.5  
9039                   MX         6.6      2014.0          950.986         4.1  
8964                   US         6.0     26244.0          934.545         7.1  
7421               JP, US         6.5    126887.0          893.650         7.4  
8948           CA, GB, US         5.9     11522.0          482.744         7.3  
9092                   CA         3.6       631.0          429.802         6.2  
9011                   FR         5.6      7782.0          389.431         6.0  
4099                   US         8.5    395600.0          356.533         8.1  
5287                   US         8.8    404920.0          352.657         8.5